### PR TITLE
[android] Fix "Can not perform this action after onSaveInstanceState"

### LINKED ISSUE #4549: [android] Can not perform this action after onSaveInstanceState
A new regression after merging #4174:

```
java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState

     FATAL EXCEPTION: Thread-2
Process: app.organicmaps.debug, PID: 9950
java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
	at androidx.fragment.app.FragmentManager.checkStateLoss(FragmentManager.java:1610)
	at androidx.fragment.app.FragmentManager.ensureExecReady(FragmentManager.java:1711)
	at androidx.fragment.app.FragmentManager.execSingleAction(FragmentManager.java:1725)
	at androidx.fragment.app.BackStackRecord.commitNow(BackStackRecord.java:317)
	at app.organicmaps.widget.placepage.PlacePageController.removePlacePageFragments(PlacePageController.java:397)
	at app.organicmaps.widget.placepage.PlacePageController.onHiddenInternal(PlacePageController.java:183)
	at app.organicmaps.widget.placepage.PlacePageController.-$$Nest$monHiddenInternal(Unknown Source:0)
	at app.organicmaps.widget.placepage.PlacePageController$1.onSheetHidden(PlacePageController.java:100)
	at app.organicmaps.widget.placepage.DefaultBottomSheetCallback.onStateChanged(DefaultBottomSheetCallback.java:33)
	at com.google.android.material.bottomsheet.BottomSheetBehavior.setStateInternal(BottomSheetBehavior.java:1388)
	at com.google.android.material.bottomsheet.BottomSheetBehavior$StateSettlingTracker$1.run(BottomSheetBehavior.java:1984)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
	at android.view.Choreographer.doCallbacks(Choreographer.java:761)
	at android.view.Choreographer.doFrame(Choreographer.java:693)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:10)
	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:7)
	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:1)
	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadForAtLeast(UiControllerImpl.java:7)
	at androidx.test.espresso.action.Tap$1.sendTap(Tap.java:4)
	at androidx.test.espresso.action.GeneralClickAction.perform(GeneralClickAction.java:4)
	at androidx.test.espresso.ViewInteraction$SingleExecutionViewAction.perform(ViewInteraction.java:2)
	at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:23)
	at androidx.test.espresso.ViewInteraction.-$$Nest$mdoPerform(Unknown Source:0)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:6)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:193)
	at android.app.ActivityThread.main(ActivityThread.java:6669)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

MediumPhone (ARM), Virtual, API level 28

https://console.firebase.google.com/u/0/project/omapsapp/testlab/histories/bh.7a38003b1edbc7b5/matrices/5196180644321985195/executions/bs.dca7c31e2afabb3a/issues

https://user-images.githubusercontent.com/170263/220216433-e9e26d55-3d8b-4f2b-8ef9-87ec2e10037d.mp4

The relevant log just before the crash:
```
I/ActivityManager(3304): Displayed app.organicmaps.debug/app.organicmaps.bookmarks.BookmarkCategoriesActivity: +157ms
02-19 21:24:53.739: I/GoogleInputMethod(7441): onFinishInput() : Dummy InputConnection bound
02-19 21:24:53.739: I/GoogleInputMethod(7441): onStartInput() : Dummy InputConnection bound
02-19 21:24:53.740: D/EGL_emulation(9950): eglMakeCurrent: 0x7104e4ca40: ver 3 0 (tinfo 0x7104fd1e60)
02-19 21:24:53.758: D/EGL_emulation(9950): eglMakeCurrent: 0x7104e4ca40: ver 3 0 (tinfo 0x7104fd1e60)
02-19 21:24:53.760: D/MapFragment(9950): MapFragment.java:68 surfaceDestroyed(): surfaceDestroyed
02-19 21:24:53.760: D/Map(9950): Map.java:188 onSurfaceDestroyed(): mSurfaceCreated = true, mSurfaceAttached = true, isAdded = true
02-19 21:24:53.760: I/OMcore(9950): organicmaps/Framework.cpp:285 DetachSurface(): Detach surface started. destroySurface = 1
02-19 21:24:53.760: I/OMcore(9950): organicmaps/Framework.cpp:298 DetachSurface(): Destroy surface.
02-19 21:24:53.784: I/OMcore(9950): drape_frontend/frontend_renderer.cpp:2276 OnContextDestroy(): On context destroy.
02-19 21:24:53.803: I/OMcore(9950): drape_frontend/backend_renderer.cpp:660 OnContextDestroy(): On context destroy.
02-19 21:24:53.826: D/OMcore(9950): drape_frontend/drape_engine.cpp:396 SetRenderingDisabled(): Rendering disabled
02-19 21:24:53.831: I/OMcore(9950): organicmaps/Framework.cpp:320 DetachSurface(): Detach surface finished.
02-19 21:24:53.836: D/EGL_emulation(9950): eglMakeCurrent: 0x7104e4ca40: ver 3 0 (tinfo 0x7104fd1e60)
02-19 21:24:53.840: D/AppBackgroundTracker(9950): AppBackgroundTracker.java:77 onActivityStopped(): onActivityStopped activity = app.organicmaps.MwmActivity@3ffae33
02-19 21:24:53.841: D/MapFragment(9950): MapFragment.java:112 onStop(): onStop
 I/BaseActivityDelegate(9950): BaseActivityDelegate.java:97 logLifecycleMethod(): MwmActivity: onStop() activity: app.organicmaps.MwmActivity@3ffae33
02-19 21:24:53.841: I/OMcore(9950): organicmaps/Framework.cpp:903 Java_app_organicmaps_Framework_nativeRemovePlacePageActivationListener(): Remove global map object listener
02-19 21:24:53.841: D/LocationHelper(9950): LocationHelper.java:420 removeListener(): listener: app.organicmaps.MwmActivity@3ffae33 count was: 2
02-19 21:24:53.841: D/LocationHelper(9950): LocationHelper.java:594 detach(): activity = app.organicmaps.MwmActivity@3ffae33
02-19 21:24:53.841: D/LifecycleMonitor(9950): Lifecycle status change: app.organicmaps.MwmActivity@3ffae33 in: STOPPED
02-19 21:24:53.842: D/AppBackgroundTracker(9950): AppBackgroundTracker.java:109 onActivitySaveInstanceState(): onActivitySaveInstanceState activity = app.organicmaps.MwmActivity@3ffae33
02-19 21:24:53.842: D/DefaultBottomSheetCallback(9950): DefaultBottomSheetCallback.java:25 onStateChanged(): State change, new = HIDDEN
```

### PR DESCRIPTION
**Pull request by rtsisyk**
Fixes #4549

### CODE REVIEW

**Comment by arnaudvergnet** [comment]
This implementation is way clearer! What I don't understand is how can this code run if the activity is destroyed? The controller is part of the activity, shouldn't it get destroyed along with the activity?

File: android/src/app/organicmaps/widget/placepage/PlacePageController.java [hunk 1]
```diff
@@ -76,6 +78,35 @@ public class PlacePageController implements Initializable<Activity>,
...
+      if (PlacePageUtils.isSettlingState(newState) || PlacePageUtils.isDraggingState(newState) ||
+          mMwmActivity.isFinishing() || mMwmActivity.isDestroyed())
+        return;
+
+      PlacePageUtils.moveViewportUp(mPlacePage, mViewportMinHeight);
+
+      if (PlacePageUtils.isHiddenState(newState))
+      {
+        onHiddenInternal();
+      }
```
Comment (biodranik) [inline android/src/app/organicmaps/widget/placepage/PlacePageController.java:96]:
```suggestion
        onHiddenInternal();
```

File: android/src/app/organicmaps/widget/placepage/PlacePageController.java [hunk 2]
```diff
@@ -76,6 +78,35 @@ public class PlacePageController implements Initializable<Activity>,
...
+    }
+
+    @Override
+    public void onSlide(@NonNull View bottomSheet, float slideOffset)
```
Comment (biodranik) [inline android/src/app/organicmaps/widget/placepage/PlacePageController.java:100]: What causes the call of this callback when activity is destroyed? Can we unsubscribe from it/cancel it when the app is closed, and avoid its call at all?

Comment (biodranik) [reply android/src/app/organicmaps/widget/placepage/PlacePageController.java:100]: @rtsisyk it was closed almost a year ago. Do we use the latest dependency version with that fix?

**Review by vng — APPROVED** [review]

File: android/src/app/organicmaps/widget/placepage/PlacePageController.java [hunk 3]
```diff
@@ -84,7 +84,8 @@ class DefaultBottomSheetCallback extends BottomSheetBehavior.BottomSheetCallback
     public void onStateChanged(@NonNull View bottomSheet, int newState)
     {
       Logger.d(TAG, "State change, new = " + PlacePageUtils.toString(newState));
-      if (PlacePageUtils.isSettlingState(newState) || PlacePageUtils.isDraggingState(newState))
+      if (PlacePageUtils.isSettlingState(newState) || PlacePageUtils.isDraggingState(newState) ||
+          mMwmActivity.isFinishing() || mMwmActivity.isDestroyed())
```
Comment (biodranik) [inline android/src/app/organicmaps/widget/placepage/PlacePageController.java:88]: Will it help to debug/fix a bug of unnecessary calling `onStateChanged()` if we log these isFinishing and isDestroyed states?

File: android/src/app/organicmaps/widget/placepage/PlacePageController.java [hunk 4]
```diff
@@ -96,6 +97,8 @@ public void onStateChanged(@NonNull View bottomSheet, int newState)
     @Override
     public void onSlide(@NonNull View bottomSheet, float slideOffset)
     {
+      if (mMwmActivity.isFinishing() || mMwmActivity.isDestroyed())
```
Comment (biodranik) [inline android/src/app/organicmaps/widget/placepage/PlacePageController.java:100]: Let's log this case as an error (does logging error crashes/asserts now in debug? C++ code asserts.)

File: android/src/app/organicmaps/widget/placepage/PlacePageController.java (hunk 3 above)
Comment (biodranik) [reply android/src/app/organicmaps/widget/placepage/PlacePageController.java:88]: E.g. to crash/assert in debug.

Comment (biodranik) [reply android/src/app/organicmaps/widget/placepage/PlacePageController.java:88]: Isn't onStateChanged a BottomSheet callback that is called on any state change, not only after onSaveInstanceState?

**Pull request merged**